### PR TITLE
O3-2908 List Display Delay on Bill Form During Search

### PR DESCRIPTION
**Pull request by CynthiaKamau**
## Requirements
- [ ] This PR has a title that briefly describes the work done including a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) type prefix and a Jira ticket number if applicable. See existing PR titles for inspiration.

#### For changes to apps
- [ ]  My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).

#### If applicable
- [ ] My work includes tests or is validated by existing tests.

## Summary

Details are captured [here](https://openmrs.atlassian.net/browse/O3-2908%20List%20Display%20Delay%20on%20Bill%20Form%20During%20Search)

## Screenshots

https://github.com/openmrs/openmrs-esm-billing-app/assets/19533785/11ae9176-0c46-48ff-9bfa-a598188d1553

### CODE REVIEW

File: src/billing-form/billing-form.component.tsx [hunk 1]
```diff
@@ -117,8 +119,9 @@ const BillingForm: React.FC<BillingFormProps> = ({ patientUuid, closeWorkspace }
 
   const { data, error, isLoading, isValidating } = useFetchSearchResults(searchVal, category);
 
-  const filterItems = (val) => {
+  const filterItems = async (val) => {
```
Comment (kajambiya) [inline src/billing-form/billing-form.component.tsx:122]: Hi @CynthiaKamau 
I've gone through the code and all looks good but I was wondering why the `filterItems` function was made asynchronous. Given the logic, it looks like an overkill to me.

File: src/billing-form/billing-form.component.tsx [hunk 2]
```diff
@@ -207,22 +224,22 @@ const BillingForm: React.FC<BillingFormProps> = ({ patientUuid, closeWorkspace }
...
+        <RadioButton labelText={t('stockItem', 'Stock Item')} value="Stock Item" id="stockItem" />
+        <RadioButton labelText={t('service', 'Service')} value="Service" id="service" />
       </RadioButtonGroup>
 
       <div>
         <Search
           id="searchField"
           size="lg"
-          placeholder="Find your drugs here..."
+          placeholder="Find your items here..."
```
Comment (pirupius) [inline src/billing-form/billing-form.component.tsx:235]: Let's pass the string in a translator

File: src/billing-form/billing-form.component.tsx [hunk 3]
```diff
@@ -149,9 +155,20 @@ const BillingForm: React.FC<BillingFormProps> = ({ patientUuid, closeWorkspace }
       });
 
       setSearchOptions(options.filter((option) => option)); // Filter out undefined/null values
+      if (options.length === 0 || error) {
+        setNoResultsMessage('No results found.');
+      } else {
+        setNoResultsMessage('');
+      }
+    } else {
+      setNoResultsMessage('No results found.');
```
Comment (pirupius) [inline src/billing-form/billing-form.component.tsx:164]: This feels redundant. We can add a string if there no results or else there's no need for the extra state to hold am empty string

File: src/billing-form/billing-form.component.tsx (hunk 1 above)
Comment (pirupius) [inline src/billing-form/billing-form.component.tsx:122]: From the look of things this is being passed to the `debouncedFilterItems` with extra steps of filtering by service type. Isn't it doing too much as is and causing confusion?

We could eliminate some extra steps and let the `debounce` have the search logic and leave the filter to only do filtering.

File: src/billing-form/billing-form.component.tsx (hunk 3 above)
Comment (CynthiaKamau) [reply src/billing-form/billing-form.component.tsx:164]: The backend returns an empty array and the user continues doing the search not knowing that there is nothing to search through, thats why i added the message

File: src/billing-form/billing-form.component.tsx (hunk 1 above)
Comment (CynthiaKamau) [reply src/billing-form/billing-form.component.tsx:122]: Let me try consolidate everything in `debounce`

File: src/billing-form/billing-form.component.tsx [hunk 4]
```diff
@@ -207,36 +211,38 @@ const BillingForm: React.FC<BillingFormProps> = ({ patientUuid, closeWorkspace }
...
           disabled
-          closeButtonLabelText="Clear search input"
-          onChange={() => {}}
+          closeButtonLabelText={t('clearSearchInput', 'Clear search input')}
           className={styles.billingItem}
-          onKeyUp={(e) => {
-            filterItems(e.target.value);
-          }}
+          placeholder={t('findYourItemsHere', 'Find your items here')}
+          labelText={t('findYourItemsHere', 'Find your items here')}
```
Comment (pirupius) [inline src/billing-form/billing-form.component.tsx:226]: What do u think of a more direct branding like so;

```suggestion
          placeholder={t('searchItems', 'Search items and services')}
          labelText={t('searchItems', 'Search items and services')}
```

File: src/billing-form/billing-form.component.tsx [hunk 5]
```diff
@@ -37,6 +40,8 @@ const BillingForm: React.FC<BillingFormProps> = ({ patientUuid, closeWorkspace }
   const [category, setCategory] = useState('');
   const [saveDisabled, setSaveDisabled] = useState<boolean>(false);
   const [addedItems, setAddedItems] = useState([]);
+  const [searchTerm, setSearchTerm] = useState('');
+  const debouncedSearchTerm = useDebounce(searchTerm);
```
Comment (pirupius) [inline src/billing-form/billing-form.component.tsx:44]: 🚀

**Review by pirupius — APPROVED** [review]

**Pull request merged**